xapian-core/cluster/cluster.cc

```diff
@@ -644,12 +637,17 @@ StemStopper::add(const string &term)
 {
     LOGCALL_VOID(API, "StemStopper::add", term);
     const string &stemmed_term = 'Z' + stemmer(term);
-    if (stem_action == STEM_NONE)
-	stop_words.insert(term);
-    else if (stem_action == STEM_ALL_Z)
-	stop_words.insert(stemmed_term);
-    else {
-	stop_words.insert(term);
-	stop_words.insert(stemmed_term);
+    switch (stem_action) {
+	case STEM_NONE:
+	    stop_words.insert(term);
+	    break;
+	case STEM_ALL_Z:
+	case STEM_ALL:
```

Comment:
And we shouldn't call the stemmer unless we're going to actually use the result, as I believe I already said.  The stemming algorithm is pluggable by the user, so could be arbitrary expensive.  Even the built-in stemming algorithms generally do a non-trivial amount of work.